File: backend/domain/entities/product.py
"""
Product entity - Products for each business owner
"""
from datetime import datetime
from typing import Optional, List, Any
from dataclasses import dataclass
from decimal import Decimal


@dataclass
class Product:
    """
    Product entity representing products in inventory.
    
    Business Rules:
    - product_code unique per owner
    - base_unit_id required for inventory tracking
    - base_price must be >= 0
    - cost_price used for profit calculation
    """
    id: Optional[int]
    owner_id: int
    product_code: str
    name: str
    base_unit_id: int
    base_price: Decimal
    category: Optional[str] = None
    description: Optional[str] = None
    cost_price: Optional[Decimal] = None
    barcode: Optional[str] = None
    image_url: Optional[str] = None
    is_active: bool = True
    available_quantity: Optional[Decimal] = None
    units: Optional[List[Any]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    def __post_init__(self):
        """Validate product data"""
        if not self.product_code or len(self.product_code.strip()) == 0:
            raise ValueError("Product code is required")
        
        if not self.name or len(self.name.strip()) == 0:
            raise ValueError("Product name is required")
        
        if self.base_price < 0:
            raise ValueError("Base price cannot be negative")
        
        if self.cost_price is not None and self.cost_price < 0:
            raise ValueError("Cost price cannot be negative")
    
    def calculate_profit_margin(self) -> Optional[Decimal]:
        """
        Calculate profit margin percentage.
        Returns None if cost_price not set.
        """
        if self.cost_price is None or self.cost_price == 0:
            return None
        
        profit = self.base_price - self.cost_price
        margin = (profit / self.cost_price) * 100
        return round(margin, 2)
    
    def is_profitable(self) -> bool:
        """Check if product is profitable"""
        if self.cost_price is None:
            return True  # Unknown, assume profitable
        return self.base_price > self.cost_price
    
    def get_profit_per_unit(self) -> Decimal:
        """Get profit per unit in base price"""
        if self.cost_price is None:
            return Decimal("0")
        return self.base_price - self.cost_price

File: frontend/mobile/lib/models/product.dart
class Product {
  final int id;
  final String productCode;
  final String name;
  final String? description;
  final String? category;
  final double price;
  final double? costPrice;
  final double? availableQuantity;
  final String? imageUrl;
  final List<ProductUnit>? units;

  Product({
    required this.id,
    required this.productCode,
    required this.name,
    this.description,
    this.category,
    required this.price,
    this.costPrice,
    this.availableQuantity,
    this.imageUrl,
    this.units,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: _parseInt(json['id']),
      productCode: json['product_code'] ?? '',
      name: json['name'] ?? '',
      description: json['description'],
      category: json['category'],
      price: _parseDouble(json['base_price'] ?? json['price']),
      costPrice: _parseDouble(json['cost_price']),
      availableQuantity: _parseDouble(json['available_quantity']),
      imageUrl: json['image_url'],
      units: json['units'] != null
          ? (json['units'] as List).map((i) => ProductUnit.fromJson(i)).toList()
          : null,
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }
}

class ProductUnit {
  final int id;
  final String name;
  final double conversionRate;
  final double price;

  ProductUnit({
    required this.id,
    required this.name,
    required this.conversionRate,
    required this.price,
  });

  factory ProductUnit.fromJson(Map<String, dynamic> json) {
    return ProductUnit(
      id: _parseInt(json['id']),
      name: json['unit_name'] ?? json['name'] ?? '',
      conversionRate: _parseDouble(json['conversion_rate'] ?? 1.0),
      price: _parseDouble(json['price']),
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }
}
